<h1>Урок 1. Алгоритм линейной регрессии. Градиентный спуск.</h1>

<h2>1. Подберите скорость обучения (eta) и количество итераций</h2>

In [1]:
import numpy as np

In [2]:
X = np.array([[ 1,  1],
              [ 1,  1],
              [ 1,  2],
              [ 1,  5],
              [ 1,  3],
              [ 1,  0],
              [ 1,  5],
              [ 1, 10],
              [ 1,  1],
              [ 1,  2]])
y = [45, 55, 50, 55, 60, 35, 75, 80, 50, 60]

In [3]:
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

In [4]:
# Параметры:
eta = 1e-2
n_iter = 443

n = X.shape[1]
err = None

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
        \nNumber of iterations: {n_iter}\
        \nLearning rate = {eta} \
        \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    for k in range(W.shape[0]):
        W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    if i % 10 == 0:
        eta /= 1.1

print(f'Iteration #{n_iter}: W_new = {W}, MSE = {round(err, 2)}')

Number of objects = 2         
Number of iterations: 443        
Learning rate = 0.01         
Initial weights = [1.  0.5] 

Iteration #443: W_new = [44.63509213  3.89000383], MSE = 44.05


<b>при n_iter = 0.01 достаточно 443 итерации для получения результата с MSE = 44.05</b>

<h2>2*. В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее</h2>

In [5]:
n = X.shape[1]

eta = 1e-2
n_iter = 100

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    #
    W -= eta * (1/n * 2 * np.dot(X[:, 0], y_pred - y))
    #
    if i % 10 == 0:
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err,2)}')

Number of objects = 2        
Learning rate = 0.01        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [6.4 5.9], MSE = 3047.75
Iteration #10: W_new = [14.4510224 13.9510224], MSE = 861.1
Iteration #20: W_new = [14.49970385 13.99970385], MSE = 874.17
Iteration #30: W_new = [14.49999821 13.99999821], MSE = 874.25
Iteration #40: W_new = [14.49999999 13.99999999], MSE = 874.25
Iteration #50: W_new = [14.5 14. ], MSE = 874.25
Iteration #60: W_new = [14.5 14. ], MSE = 874.25
Iteration #70: W_new = [14.5 14. ], MSE = 874.25
Iteration #80: W_new = [14.5 14. ], MSE = 874.25
Iteration #90: W_new = [14.5 14. ], MSE = 874.25


<b>W -= eta * (1/n * 2 * np.dot(X[:, 0], y_pred - y))</b>

<h2>3*. Вместо того, чтобы задавать количество итераций, задайте другое условие останова алгоритма - когда веса перестают изменяться меньше определенного порога 𝜖 </h2>

In [6]:
n = X.shape[1]

eta = 1e-2
n_iter = 100

#
e_krit = 1e-5
W_new = np.array([1, 0.5])
W = np.array([1000, 0.5])
#
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')
#
i = 0
while np.abs(W[0] - W_new[0]) > e_krit:
#
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    #
    W = W_new.copy()
    W_new -= eta * (1/n * 2 * np.dot(X[:, 0], y_pred - y))
    #
    if i % 10 == 0:
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err,2)}')
    #
    i += 1
    #

print(f'Iteration #{i}: W_new = {W}, MSE = {round(err,2)}')

Number of objects = 2        
Learning rate = 0.01        
Initial weights = [1.e+03 5.e-01] 

Iteration #0: W_new = [1.  0.5], MSE = 893156.75
Iteration #10: W_new = [13.55824 13.05824], MSE = 1116.12
Iteration #20: W_new = [14.48303242 13.98303242], MSE = 874.2
Iteration #30: W_new = [14.49976754 13.99976754], MSE = 874.23
Iteration #35: W_new = [14.5000203 14.0000203], MSE = 874.25


<b>
W = W_new.copy()


W_new -= eta * (1/n * 2 * np.dot(X[:, 0], y_pred - y))
</b>